feat: adds CommonMetadata to HelmRelease postrenderer

abhijith-darshan opened the pull request:
This PR addresses Issue #1217 to pass common labels and common annotations to `postRender` kustomize

Following the existing convention of `CommonMetadata` that is used in `kustomize-controller` to add labels and annotations, `CommonMetadata` is added to `HelmRelease` spec so that a common set of annotations and labels can be propagated to all resources in the helm chart without having to know the values schema of the chart

abhijith-darshan commented:
@hiddeco any further recommendations for this PR?

hiddeco commented on api/v2/helmrelease_types.go:
```diff
@@ -185,6 +185,14 @@ type ValuesReference = meta.ValuesReference
+	// Common labels to set on all resources.
+	// When duplicate labels are found across postRenderers, the last one wins.
+	// +optional
+	CommonLabels map[string]string `json:"commonLabels,omitempty"`
```
I wonder if it might be better to expose the full options available in Kustomize, rather than giving the false impression this makes use of the deprecated `commonLabels` feature.

abhijith-darshan replied:
I saw this PR - https://github.com/fluxcd/kustomize-controller/pull/817

> The main difference to the Kustomize [commonLabels](https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/commonlabels/) and [commonAnnotations](https://kubectl.docs.kubernetes.io/references/kustomize/kustomization/commonannotations/), is that the controller sets the labels and annotations only to the top level metadata field, without patching the Kubernetes Deployment spec.template or the Service spec.selector.

but in this case it would be useful to propagate the information to even the annotations or labels of the pods created by `Deployment` 

As this is not the use case of flux kustomize resource I would request to allow this option in the helm postRenderer.

How about I change this to commonMetadata format as suggested but expose further boolean option to `IncludeTemplates`?

Would that be acceptable?

stefanprodan replied:
Changing the label selectors will always result in a Helm upgrade failure with an immutable field error. I don't see why would we expose a way for people to shoot themselves in the foot.

abhijith-darshan replied:
@stefanprodan - sorry I don't mean Deployment's `.spec.selector.matchLabels` and you are right they are immutable...

I meant Deployment's `template.metadata.labels` which can be modified and that is what kustomize `IncludeTemplates` does.

Which is what I did -

```go
	if k.CommonLabels != nil {
		cfg.Labels = make([]kustypes.Label, 0)
		cfg.Labels = append(cfg.Labels, kustypes.Label{
			Pairs:            k.CommonLabels,
			IncludeSelectors: false,
			IncludeTemplates: true,
		})
	}
```

abhijith-darshan commented:
> @abhijith-darshan the `commonMetadata` should be added to `.spec`, same as we do for Flux Kustomization. Please copy the type instead of adding a dependency to kustomize-controller. As for the implementation, we already have code that adds labels to all resources to track the HelmRelease owner, we can extend that code to add the extra metadata.

As suggested - 

- `commonMetadata` from `.spec.postRenderers` to `.spec`.
- removed the dependency to `kustomize-controller` and copied over the types for `CommonMetadata`
- merged the logic of `origin_labels.go` into `common_renderer.go` 
- introduced `transformers.go` to re-use _kustomize builtins_ `LabelTransformer` and `AnnotationsTransformer`
- `BuildPostRenderers` gives precedence to `OriginLabels` rendering over `CommonMetadata` in case of label key collision (unlikely to happen but the edge case is covered)
- Updated tests but please let me know if I have missed something

stefanprodan reviewed (commented):
Please add a section to the docs about common metadata before https://github.com/fluxcd/helm-controller/blob/main/docs/spec/v2/helmreleases.md#post-renderers

stefanprodan commented on api/go.mod:
```diff
@@ -21,9 +21,9 @@ require (
-	golang.org/x/net v0.39.0 // indirect
-	golang.org/x/text v0.24.0 // indirect
-	golang.org/x/tools v0.32.0 // indirect
+	golang.org/x/net v0.40.0 // indirect
```
Please undo all changes to go.mod and go.sum in the api and the main ones

abhijith-darshan replied:
done in [132abe9](https://github.com/fluxcd/helm-controller/pull/1223/commits/132abe998801a8b45128f364326c4be68e81a9a7)

abhijith-darshan commented on internal/postrender/transformer.go:
```diff
@@ -0,0 +1,55 @@
+package postrender
```
Done in [8576fb5](https://github.com/fluxcd/helm-controller/pull/1223/commits/8576fb56608829bc648258dc9d90dfa0f2e73424)

stefanprodan commented:
@abhijith-darshan can you please squash the commits
